text/0000-pack-package-refactor.md

```diff
@@ -0,0 +1,58 @@
+# Meta
+[meta]: #meta
+- Name: (fill in the feature name: My Feature)
+- Start Date: 02/24/21
+- Author(s): dwillist
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a restructuring of the `pack` codebase into reusable modular packages.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+    - Maintaining a modular and re-usable codebase enables contributors to be more effective, and encourages new adopters. 
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC is looking accomplish the following:
+- enable pack library consumers by moving code out of `internal`
+- ease the process of contribution by having well defined connectable components.
+- Follow a general process of if data is externally visible on a pack produced artifact. We should provide structs for decoding.
+
+To accomplish the above we propose the following package structure:
+
+```
+.
+├── cmd/
+│   ├── pack/ → current pack cli
+│   └── fetcher/ → cli to download buildpacks
+├── pkg/
+│   ├── archive/ → lib for converting multiple sources into OCI layers
+│   ├── client/ → current pack client library
+│   ├── descriptor/ → lib for processing project descriptor
+│   ├── inspect/ → lib for inspecting various CNB resources (app image, buildpack, builder, stack, ...)
+│   ├── internal/ → all non-exported go code
+│   ├── stack/ → lib for mixin validation & stack labels
+│   ├── stack/internal/ → keep internal details close to exported APIs.
+│   ├── builder/ → lib for builder labels & builder manipulation
+│   ├── builder/internal/ → ...
+│   ├── buildpackage/ → lib for buildpackage label & buildpack extraction from buildpackages
+│   ├── buildpackage/internal/ → ...
+│   └── resolve/ → lib for resolving buildpacks via Buildpacks URIs
```

Comment:
Given a lifecycle `prepare` phase may eventually want to use some of this logic (resolving buildpack uris, pulling buildpacks) and we don't want the lifecycle to import pack, should we think about creating a new repo (`distribution`?) for these packages?

We could do this later but it might make sense to do know, while we are refactoring to avoid problems disentangling later (if for example internal packages are used by packages that should be moved and also by packages that should remain in this repo).

Reply:
Part of me is hesitant to move stuff into a new repository at this moment because we don't yet have the right seams and am afraid that the effort would prevent us from moving forward to even a slightly better structure.

In my mind, this RFC can be seen as the first step into potentially moving stuff out. I would think of it as a "promotion" strategy for any logic that exists in pack that we may want to redistribute.

ie:
`/pack/archive` -> `imgutil/archive`
`/pack/pkg/resolve` -> `<TBD>/resolve`

Reply:
The makes sense. We can promote as a first step and think about extraction later if it is warranted.